src/client_side_reply.cc

```diff
@@ -910,15 +910,22 @@ purgeEntriesByUrl(HttpRequest * req, const char *url)
 
     for (HttpRequestMethod m(Http::METHOD_NONE); m != Http::METHOD_ENUM_END; ++m) {
         if (m.respMaybeCacheable()) {
-            if (StoreEntry *entry = storeGetPublic(url, m)) {
-                debugs(88, 5, "purging " << *entry << ' ' << m << ' ' << url);
+            const cache_key *key = storeKeyPublic(url, m);
+            debugs(88, 5, m << ' ' << url << ' ' << storeKeyText(key));
+            if (StoreEntry *entry = Store::Root().get(key)) {
+                entry->lock("purgeEntriesByUrl");
 #if USE_HTCP
                 neighborsHtcpClear(entry, url, req, m, HTCP_CLR_INVALIDATION);
                 if (m == Http::METHOD_GET || m == Http::METHOD_HEAD) {
                     get_or_head_sent = true;
                 }
 #endif
-                entry->release();
+                // entry->release() notifies waiting workers, including this one
+                entry->release(true);
+
+                entry->unlock("purgeEntriesByUrl");
+            } else {
+                Store::Root().unlinkByKeyIfFound(key); // does not broadcast but there were no waiting workers
```

Comment:
This comment is still misleading or even wrong. There could be workers working on an entry with `key`, of course. What we are trying to say here is that we should not worry about ignoring them. Why? I believe we can safely ignore such workers because either the entry has already been marked for deletion (and so another notification would make no difference) or Controller::find() failed to create new or join an existing Transients entry (and so we would not be able to notify anybody anyway). If my logic is correct, then let's replace this comment with something like this:

```
// The call below does not notify workers but that is OK here because nil
// Store::Root().get() guarantees that either the matching entry has already
// been marked (so new notifications are useless) OR get() could not create
// (or join) a Transients entry (so notifications are impossible).
```

Reply:
```
// The call below does not notify workers but that is OK here because nil
// Store::Root().get() guarantees that either the matching entry has already
// been marked (so new notifications are useless) OR get() could not create
// (or join) a Transients entry (so notifications are impossible).
```

I do not mind this in general, however note that if get() returns nil that only means that "it could join a Transients entry" (as well as any other of shared or local storages), we don't know whether the entry was deleted or have not been created. Also, AFAIK, get() does not create Transients entries: it is done by Controller::allowCollapsing() and Transients::startWriting().

Reply:
> if get() returns nil that only means that "it could join a Transients entry"

I assume you meant "could _not_ join a Transients entry".

> (as well as any other of shared or local storages)

Right, but those other stores do not matter as far as notifications are concerned, and the comment we are discussing is about notifications.

> get() does not create Transients entries

Good point! It only creates StoreEntries from existing Transient entries. 

Another problem with that comment is that it is actually possible to broadcast a notification even if we could not join an existing Transients record -- broadcast messages just need `xitTable.index` value which we could compute. Such broadcasts would be useless because if we could not join then nobody else could, and so there are no readers to broadcast to (and the Transients writer does not care about the delete broadcasts, right?).

Here is a fixed(?) comment:

```
// The call below does not notify workers but that is OK here because nil
// Store::Root().get() guarantees that the matching Transients entry does
// not exist (so there are no readers to notify), could not be joined yet
// (ditto), or has already been marked (so new notifications are useless).
```

However, all the conditions in the above comment were true only at the get() time. They may no longer be true at the unlinkByKeyIfFound() time! What happens if a new reader comes and manages to attach to the Transients entry after our failed get() but before our unlinkByKeyIfFound() call? Will that reader have to wait for the next store write to realize that it can never attach to the stores and has to abandon the current transaction? If yes, is there a simple way to improve this? 

Perhaps unlinkByKeyIfFound() should broadcast using raw Transients index but it feels like we should just reorder the markings instead:

1. Mark stores (not Transients!) by key.
1. Root().get().
1. If we got an entry, do the usual entry->release(), with Broadcast().
1. Otherwise, do nothing.

What do you think?

Reply:
> What happens if a new reader comes and manages to attach to the Transients entry after our failed get() but before our unlinkByKeyIfFound() call? 

BTW, similar logic should be applied to the writer itself: an entry may be locked for Transients writing (without appending) at the time of get() but become 'appending' (and available for another get()) just before unlinkByKeyIfFound(). Then the skipped broadcasting would not release() that local writer entry (Controller::syncCollapsed()).

>  but it feels like we should just reorder the markings instead

What is the need of (1)? If Root.get() works for us(Transients is not marked) why it will not work for others? I.e., If shared Transients entry is unavailable at (2) but becomes available just before (3) wont we face same problems?

Reply:
> similar logic should be applied to the writer itself: an entry may be locked for Transients writing (without appending) at the time of get() but become 'appending' (and available for another get()) just before unlinkByKeyIfFound(). Then the skipped broadcasting would not release() that local writer entry (Controller::syncCollapsed()).

Agreed. The writer problem cannot be (easily) solved by broadcasting AFAICT. Instead, right before switching to appending mode, the writer must check waitingToBeFreed and release() its entry. If that code is missing, please add it. I think there is already similar code somewhere.

> What is the need of (1)?

Step 1 calls unlinkByKeyIfFound() for memory and disk stores (if any).

> If Root.get() works for us(Transients is not marked) why it will not work for others? 

Root.get() will work for others as well, but they will get our notification and will be fine. They will not attach themselves to stores because we marked those first (in step 1).

> If shared Transients entry is unavailable at (2) but becomes available just before (3) wont we face same problems?

We will! Glad you found this bug. I also realized that my step 4 was missing unlinkByKeyIfFound() for Transients, but adding that would not fix this bug.

Also, I really do not want to broadcast based on unlocked Transient index values because key:index mapping is (or will be) unreliable (if StoreMap supports hash collision resolution).

I do not have any good solutions to offer right now. Back to the drawing board... If you can offer something that addresses all these problems, please do!

Reply:
> The writer problem cannot be (easily) solved by broadcasting AFAICT. Instead, right before switching to appending mode, the writer must check waitingToBeFreed and release() its entry. If that code is missing, please add it. 

Your suggestion would work for a case when writer switches into 'appending' mode just _after_ unlinkByKeyIfFound() (i.e., when waitingToBeFreed is set already). However I was asking about another case, when writer becomes 'appending' just _before_ unlinkByKeyIfFound(): in this case waitingToBeFreed is not set and writer's local StoreEntry will not be released.

Reply:
> If you can offer something that addresses all these problems

I have a feeling that the current purgeEntriesByUrl() construction causes numerous problems which is difficult to address. How would I simplify it?
Again, purgeEntriesByUrl() purpose is:
1. Mark storages by the given cache_key in order to refuse new clients attach to them.
2. Release(or abort) all existing local StoreEntries (writer should be only released). Transients broadcasting should do this work.

So, roughly (optimization aside), within purgeEntriesByUrl() we would need the only unlinkByKeyIfFound(key), if it could broadcast, calculating sfileno by the given cache_key.

Reply:
Not quite. The *purpose* of the purge functionality is to remove objects from the usable HTTP object cache(s) - what you list as #1. Your #2 is just the cleanup that might be done afterwards, not part of the purpose itself.

Given the purpose is about objects being in the cache it does not imply aborting them when there are active readers OR writers. Aborting the writer (or just nul-routing its future data) is just an optimization when there are no readers left. Aborting readers would be wrong - they can finish reading (including new data from a writer that appears after the purge - if any).

Reply:
> Eduard: when writer becomes 'appending' just before unlinkByKeyIfFound(), waitingToBeFreed is not set and writer's local StoreEntry will not be released.

Agreed.

Reply:
@yadij description of the purgeEntriesByUrl() purpose is more accurate. Indeed, purgeEntriesByUrl() must make matching entries inaccessible to new clients. The rest is essentially desired side effects, such as eventual removal or reuse of the unusable slots.

@yadij, when @eduard-bagdasaryan is talking about reader aborts, he is (or should be) talking about aborting _unattached_ collapsed requests. Those aborts are _not_ desirable in principle, but are currently necessary to prevent stuck collapsed requests. If we can avoid them without drastic Squid SMP storage rewrites, we will. If we cannot, they are not that harmful because Squid retries failed collapsed requests automatically. Other readers are not and will not be aborted.

> Eduard: within purgeEntriesByUrl() we would need the only unlinkByKeyIfFound(key), if it could broadcast, calculating sfileno by the given cache_key.

Unfortunately, "calculating sfileno by the given cache_key" is impossible to do reliably within the current framework: As I may have mentioned earlier, the mapping from key to Transients fileno is not (or will not be) stable when we cannot lock the Transients entry, and the unlockable Transients entry is one of the cases we are trying to address here.

I believe I have identified the sources of the problems this PR is facing in this and several other places. I am working on describing them and suggesting a solution.

Reply:
I believe this PR progress is hindered by SMP workers inability to communicate with remote readers and writers of cache entries in certain states. Broadly speaking, the existing code attempts to address this problem by splitting use cases into two different categories:

- Controller::get() returns a StoreEntry (i.e., a "hit")
- Controller::get() returns nil

The split itself does not solve the problem because the second item does not eliminate the original communication need/challenges and offers no new tools to address them.

(BTW, the impact of the communication problems is made worse by long periods of time when some (not atypical) cache entries can be in a problematic write-only state where communication may currently be impossible. If those periods were nearly always very short, we could have ignored them, at least for now.)

I propose the following solution:

1. Ensure that each active cache entry has the corresponding Transients entry. Here, "active" means "being written or read" (i.e., has a writer and/or reader(s)). This guarantee applies to both SMP memory and SMP disk stores.
1. Add Controller::broadcastDelete(key) method that calls `transients->get()`, locks and release()s the returned local StoreEntry object, calls the old `Broadcast(entry)` function with that local StoreEntry object, and finally unlocks it. This new broadcastDelete() method will do nothing if Transients::get() returns nil (which includes cases where the matching Transients entry has already been marked for deletion, avoiding excessive broadcasts).
1. Adjust unlinkByKeyIfFound() methods to call broadcastDelete(key) when marking their store entry for deletion (for the first time). At most one such broadcast will be sent for a given cache entry because BroadcastDelete() will automatically block repeated attempts to broadcast deletion of the same key.

Once the above guarantees are provided and methods are adjusted, the purgeEntriesByUrl() loop iteration code would boil down to something like this:

```
if (m.respMaybeCacheable()) {
    ...
    neighborsHtcpClear(NULL, url, req, m, HTCP_CLR_INVALIDATION);
    ...
    Store::Root().unlinkByKeyIfFound(key);
}
```

I am not sure all readers must create Transients entry in item 1 of the proposed solution, but we know that collapsed readers must do that to avoid getting stuck, and I am worried about other readers deciding to swap out the locally read object if they do not know that the object was deleted by another worker.

Other code adjustments are likely to be necessary, of course, but I hope the above sketches all the major ones.

Do you think the above solution works?

Reply:
> Ensure that each active cache entry has the corresponding Transients entry. 

I see a problem when we load entry from caches (shared memory or disk). After this loading, a local StoreEntry is created. According to the requirement, we should immediately create Transients entry for this StoreEntry (using the designated Transients::startWriting() for this). However we can't do this because Transients::startWriting() requires RequestFlags and HttpRequestMethod parameters, not available neither in MemStore::get(), no in SwapDir::get(). These problems were recently discussed in another [issue](https://github.com/squid-cache/squid/pull/50). BTW, it looks like the problems with 'clobbering' URIs/method for MemStore also exist for Rock cache: an entry, loaded from Rock cache, does not know its original URIs/method and always re-writes them with mem_obj->setUris() inside ClientReplyContext::doGetMoreData().

So, in order to make this requirement working, we should create a 'half-backed' Transients entry(without filling Transients::extras), and then rely on setUris(), as other storages do.

Reply:
@eduard-bagdasaryan Just to fill couple gaps in my understanding of the issue in hands.
My understanding is that the original store code relies on specific assumptions about the stored object and due to this we this coding style, am I right?..
From my memory I remember that the "StoredID" was written inot the object but the method was part of the hash that was stored inside the store swap file and the basic reason was probably because of HTTP/1.0 did only allow specific METHOD+URL to be cached while HTTP/1.1 changed this.

Reply:
> My understanding is that the original store code relies on specific assumptions about the stored object and due to this we see this coding style

Not sure that I understood your question correctly: what style are you talking about? If you mean mentioned StoreEntry's URIs/method fields, then yes, StoreEntry relies on their original(correct) values in order to do correct processing and logging.

> From my memory I remember that the "StoredID" was written into the object but the method was part of the hash

AFAIK, both StoreID and method are parts of the cache_key: see key creation in storeKeyPublicByRequestMethod(). The problem is that this information is not loaded into MemObject's URIs/method fields when loading entries from stores.

Reply:
@eduard-bagdasaryan You got me right!
The MemObject's relies on the md5 hash of the "method+StoreID", this is one of this things which got complicated when Vary and other content ID headers appeared.

Reply:
> Transients::startWriting() requires RequestFlags and HttpRequestMethod parameters, not available neither in MemStore::get(), no in SwapDir::get().

Would making that information available be worse than creating half-baked entries?

Reply:
> Would making that information available be worse than creating half-baked entries?

How do you suggest to make this information available? We have an XXX inside MemStore::get() which should be addressed for this (as recently we discussed in PR 50). This is (as I understood) a separate, non-trivial task, and we decided to postpone it. Moreover, as I noted above, rock cache probably faces the same problem: URIs/method are not stored/restored from it.

Reply:
> How do you suggest to make this information available?

My first attempt would probably revolve around making the Store::get() parameter into a simple class and then adding the URLs/method/flags to that class as members. I do not know whether that is a correct or the best solution -- I have not studied that code to answer your question. Since this answer seems straightforward, I suspect you have already considered and rejected it, but I do not know why.

> We have an XXX inside MemStore::get() which should be addressed for this (as recently we discussed in PR 50). This is (as I understood) a separate, non-trivial task, and we decided to postpone it.

Your understanding is correct. However, that decision was for a small PR #50 that focused on fixing a very confined bug. This PR does not yet fix the targeted problem, and, IIRC, we know of no viable alternatives to creating Transients entries earlier. Thus, if that old XXX prevents us from creating those Transients entries earlier, we should consider fixing that XXX. I am not saying that fixing that XXX has to be a correct or the best solution, but it seems worth _considering_.

> Moreover, rock cache probably faces the same problem

... making the idea of changing the common get() API more appealing.

Reply:
I tried to move forward that plan.Two latest e9a21e7 and 548f9a8 revisions addressed discussed and newly found problems. In this approach, I moved the code which creates transients for each public StoreEntry from low-level hashInsert() to higher levels, primary where createStoreEntry() is used (except Storage::get() methods). A still open question is: are there all required createStoreEntry() uses covered? E.g., I believe we should not create transients for some sorts of entries, 'special' or 'cache manager-related', but there can be still cases where this code should be added.

Reply:
I tested latest adjustments with our Daft test. One problem occasionally happens: when several clients request a disk-cached entry, Squid returns unexpected "503 Service Unavailable" responses instead of "200" hits. Cache.log discovery showed that the reason is Transients entry creation collision: two workers simultaneously find the entry in Rock::SwapDir::get(), create StoreEntry objects and try to create Transients entry, locking it for writing. Unexpected 503s looks like the test imperfection, in reality we will get unexpected cache misses in this case.

Reply:
Glad our tests are capable of exposing such hidden problems!

In general, it is acceptable to return a miss when we cannot guarantee _reliable_ hit delivery. A hit test must either accommodate/tolerate/retry such misses or avoid them.

However, we should also avoid making hit delivery unreliable without a good reason. In the context of your "concurrent hits" example, I do not see such a reason (from an independent reviewer point of view; ignoring internal problems of our own making). There are two different cases to consider here:

1. A complete() store entry. AFAICT, such an entry does not need a Transients entry. There is no need to signal other workers, even when handling DELETE or similar purging events.
1. An in-complete() store entry. Such an entry must already have a Transients entry so there should be no creation conflicts at hit time.

If the above analysis is correct, then we should relax the earlier "every active Store entry must have a corresponding Transients entry" rule to "every Store entry _being written_ must have a corresponding Transients entry", right? A store reader must still attach to the existing Transients entry (if any), of course, but it should not try to create that Transients entry.

If you agree with the above analysis, please make the necessary code adjustments. Otherwise, please discuss what I have missed.

Reply:
> A complete() store entry. AFAICT, such an entry does not need a Transients entry. There is no need to signal other workers, even when handling DELETE or similar purging events.

I am thinking about possible problems with scenario:

1. There is a disk-cached entry.
2. Two clients request it at the same time, only one of them manages to create Transients entry.
3. The two clients start receiving cache hit.
4. DELETE comes on another worker, marking the the shared cache and releasing only the first StoreEntry(via notification).
5. The first client finishes getting the response and the corresponding Transients entry is freed. Our Store::Controller::markedForDeletion() protection stops working since now.
6. Another client hits the the second worker, thus accessing its local still active StoreEntry (and ignoring previous DELETE request).

Reply:
> 2. Two clients request an existing disk-cached entry at the same time, only one of them manages to create Transients entry.

With the adjusted rule, none of the reading workers will try to create a Transients entry.

> 6. Another client hits the the second worker, thus accessing its local still active StoreEntry (and ignoring previous DELETE request).

This concern is valid (and we do not even need two clients in step 2 to trigger this problem!).

I see two ways to fix this problem:

A. Go back to the original "all active store entries have Transients entries" rule and accept the increased probability of rejecting perfectly valid cache hits.

B. Stay with the reduced "store writers must create Transients entries" rule but adjust the existing `Controller::markedForDeletion()` method to return true for keys that have local StoreEntry objects reading store entries marked for deletion. Please note that such StoreEntry objects will have a lock on the store entries they are reading.

I think solution B is preferable. Does it miss any use cases?

Reply:
> Stay with the reduced "store writers must create Transients entries" rule

[be9511bb](https://github.com/squid-cache/squid/pull/46/commits/be9511bb7fe1c48a957c092cdaa5bd42eca84144) implements this suggestion. While re-running the test, the above problem has gone, but a new, similar one appeared. I found the source of that new problem and posted the [fix](https://github.com/squid-cache/squid/pull/46/commits/661a1c205a08fe8c5d9010cfdee5b2d93aedce50). After that, the test succeeded for both shared memory and disk cache.